active/0000-bicep-version-pinning.md

```diff
@@ -0,0 +1,134 @@
+```bicep
+resource otherResource 'Test.RP' = {} // <-- Bicep version "0.31.92" was used for compilation, but version "0.15.0" is required in configuration file (BCPxxx)
+```
+
+No specific change to the codebase is expected to achieve this, as Bicep already handles this today (i.e. if there are errors in a referenced module, an error is surfaced on all the references).
+
+----------------
+
+#### 2. Update Bicep compiler
+The Bicep compiler should be updated to check for the version constraint in the configuration.
+
+If no `version` constraint is specified anywhere in the `bicepconfig.json` resolution chain, the compilation should fallback to using the currently installed version.
+
+If the resolved `version` constraint does not match the locally-installed Bicep version, the compilation should fail with a clear message indicating the constraint violation. Otherwise, the compilation should continue normally using the resolved version.
+
+
+## Drawbacks
+- **Degraded visibility into registry module version constraints:** When inspecting published module source bicep file, it wouldn't be immediately clear what version was used to compile the module just by looking at the bicep file. This is because version information would be codified in the `bicepconfig.json` which is not published with the sources. However, the user can still discern the version used with the aid of the VSCode extension; the user can navigate to the compiled JSON which contains metadata about which version was used to compile the module. 
+- **Inability to specify version override for specific individual files:** Due to how `bicepconfig.json` works, it applies to all files in the directory it is placed in. Users cannot easily override the Bicep version for individual files. This limitation can potentially be restrictive in scenarios where there may be a need to compile specific individual files with different Bicep versions for compatibility or testing purposes.
+
+## Alternatives
+### Specify Bicep version in `.bicep` files:
+This might look something like the following:
+
+```bicep
+bicep.version = '0.31.92'
+
+resource test 'Test.RP' = {...}
+```
+This approach provides more granular control over the Bicep version used for each individual file, which makes it clear which version is required for a specific file, which could be useful in scenarios where different files need to be compiled with different versions. It would also be easier to discern what version was used to compile a published module by inspecting the source.
+
+However, the downside with this approach is if usage of the same version for multiple Bicep files is desired, it would require specifying the `version` in each file which can be a maintenance overhead.
+
+## Rollout plan
+
+This proposal only involves client-side changes, and will be an opt-in feature enabled when the bicep.version configuration property is specified (essentially acting as the feature flag).
+
+## Unresolved questions
```

Comment:
What happens when there's a version constraint in bicepconfig.json and the user runs Bicep CLI directly to build the file when the version doesn't match the constraint?

Reply:
In that case, the Bicep CLI should halt the build process and throw an error indicating that the version constraint is not satisfied (ideally along with the path to the bicepconfig.json). Does that align with your expectation? 
I have this under the [Update Bicep compiler section](https://github.com/Azure/bicep-reps/blob/976520790524b45d490959bdb2dc1b500fa633b7/active/0000-bicep-version-pinning.md?plain=1#L100), but can elaborate more there.

Reply:
What worries me is users who are interacting with multiple repos that may be owned by different people and pinned to different Bicep versions. Then, a single local installation of Bicep won't work very well.

Reply:
I think we need to confirm with the community to see how much they're using Bicep CLI as compared to az CLI.

Reply:
The feedback we received was most users are using Az CLI. Thinking we can prioritize our efforts there, and then revisit the Bicep CLI use-case for future?